### ISSUE TITLE
Vapor + Linux + GraphQL + RethinkDB

### ISSUE THREAD

**Issue by tunidev** [vapor/vapor#1454]
I want to create an API for my mobile app. Being an iOS developer creating my API using Swift is very appealing. I want to modern API so that's way I want GraphQL to be used. Also since my API should handle real time communication I have opted for RethinkDB. Linux should host the solution.

Any recommendation where to start ?  Tutorials ? the recommended stack ?

**Comment by calebkleveter** [comment]
Hey @tunidev, you might want to go to the [Vapor Slack](http://vapor.team/) for this kind of question. Hope you figure this out!

**Comment by twof** [comment]
Although it doesn't support the DB you're thinking of using, your best option might be to use https://github.com/graphcool/prisma

Vapor doesn't have a server side implementation of GraphQL at the moment to the best of my knowledge

**Comment by HappySwifter** [comment]
Here is graphQL implementation for swift.
https://github.com/GraphQLSwift/GraphQL

**Comment by twof** [comment]
@HappySwifter That's awesome! Excuse my lack of GraphQL knowledge, but will that work from both the client and server side?

**Comment by HappySwifter** [comment]
@twof For client side you can use this one
https://github.com/apollographql/apollo-ios

**Comment by HappySwifter** [comment]
> Has anything changed sen Feb, 2018?
> 
> I'm too looking for an awesome way to created GraphQL endpoint using Swift

You can check my graphQL vapor server example here https://github.com/HappySwifter/CapoServer

**Comment by maximveksler** [comment]
> > Has anything changed sen Feb, 2018?
> > I'm too looking for an awesome way to created GraphQL endpoint using Swift
> 
> You can check my graphQL vapor server example here https://github.com/HappySwifter/CapoServer

Thank you! That's very helpful.

You should probably add some tags there for GraphQL, Vapor and co. I couldn't find it while googling for options.

**Comment by HappySwifter** [comment]
> > > Has anything changed sen Feb, 2018?
> > > I'm too looking for an awesome way to created GraphQL endpoint using Swift
> > 
> > 
> > You can check my graphQL vapor server example here https://github.com/HappySwifter/CapoServer
> 
> Thank you! That's very helpful.
> 
> You should probably add some tags there for GraphQL, Vapor and co. I couldn't find it while googling for options.

Yes, you are right. Added some tags

**Comment by alexsteinerde** [comment]
I wrote an article about how to use Vapor with Graphiti (package based on GraphQL with syntax sugar): https://medium.com/@alexsteinerde/how-to-build-a-graphql-server-with-swift-and-vapor-ec202c2d4318?source=friends_link&sk=27e5ffcfc250069f6b9291bcada520db The article also includes a link to my example project that mirrors the Todo example of the official Vapor template.
If you have ideas on how to bring GraphQL more into the Vapor and Swift ecosystem, let me know ;)

**Comment by delannoyk** [comment]
@alexsteinerde Thanks for your article, it's great! I was wondering if you could give an example of a model with a `Parent` and/or a custom type (i.e. an enum of some kind:
```swift
final class Todo: ... {
    enum State: ... {
        case done
        case inProgress
    }

    //...
    var state: State
}
```
)
I just started looking at it yesterday but I quickly ran into some issues.

**Comment by alexsteinerde** [comment]
I'm currently experimenting with Parents/Children and enums. When I have something to share, I'll comment on this again ;)

**Comment by delannoyk** [comment]
@alexsteinerde I made some progress on that. It's out of scope of this issue but feel free to contact me if you need more info.
Here's what I have for my enums:
```
extension CaseIterable {
    static func type<Root, Context>() -> Enum<Root, Context, Self> {
        return Enum(self, allCases.map { Value($0) })
    }
}
```
then when creating the schema, I have to add the type:
```
Schema([
    MyEnum.type()
])
```

I also have an extension for `Parent`, `Children` and `Siblings` but they are quite rough and were quite difficult to resonate about. Let me know if you'd like to get the code.

**Comment by alexsteinerde** [comment]
@delannoyk That looks great. I also explored enums in the last couple of days. And I think by extending the `Enum` type itself would be the more compelling way. I added an extension to the GraphQLKit package: https://github.com/alexsteinerde/graphql-kit/blob/master/Sources/GraphQLKit/Graphiti%2BEnum.swift
I also wrote a short article about that: https://medium.com/@alexsteinerde/using-enums-with-graphql-and-swift-c218fc6631f8?sk=b8823e631e2b32e37f007c42d3b58c0c

Regarding the Parent and Children I figured out some extension to the `Field` type. I'll push them to the package as an idea. Feel free to comment on them and create a PR if you like. I haven't worked on `Siblings` yet. 

One point that was mentioned in the Vapor Discord is the N+1 problem that would occur with parents and children. `DataLoader` would solve the problem in some kind: https://medium.com/@__xuorig__/the-graphql-dataloader-pattern-visualized-3064a00f319f
